kernel/src/user_program/elf_loader.rs

```diff
@@ -102,6 +118,61 @@ fn verify_elf_header(header: &Elf32Ehdr) -> Result<(), ElfError> {
     Ok(())
 }
 
+#[allow(unused)]
+fn is_kernel_vaddr(vaddr: *const ()) -> bool {
+    unsafe { vaddr >= PHYS_BASE }
+}
+
+#[allow(unused)]
+fn is_user_vaddr(vaddr: *const ()) -> bool {
+    unsafe { vaddr < PHYS_BASE }
+}
+
+fn validate_segment(phdr: &Elf32Phdr, file_data: &[u8]) -> Result<(), ElfSegmentError> {
+    // p_offset and p_vaddr must have the same page offset.
+    if (phdr.p_offset as usize & PGMASK) != (phdr.p_vaddr as usize & PGMASK) {
+        return Err(ElfSegmentError::DifferentPageOffset);
+    }
+
+    // p_offset must point within FILE.
+    if phdr.p_offset as usize > file_data.len() {
```

Comment:
```suggestion
    // The range must be within FILE.
    if (phdr.p_offset + phdr.p_filesz) as usize > file_data.len() {
```

I think we also need to ensure that `p_filesz` doesn't cause the end of the range to extend beyond the end of the data.

Reply:
Changed in commit e05b3a76f46eedd171bb8c2bba05a9b56ab9b4dc, could you check if it is good? thanks.

Reply:
I am thinking to change the naming of ELFSegmentError:VmRegionWrapAround to ELFSegmentError:SegmentOverflow, do you think this would be more accurate?

Reply:
Good idea with the use of `checked_add`. What about renaming `OffsetOutOfRange` to `ContentsOutOfRange` and using it both when overflow occurs, and when the result doesn't overflow but is bigger than `file_data.len()`?

Reply:
Sure

Reply:
Changed in commit 6abe23e161957c47adfdb991cf3bf299b4b74471

Reply:
Oh sorry, forgot that `VMRegionWrapAround` was used below too. I don't think we should use `ContentsOutOfRange` there cause that doesn't really describe what went wrong in that case. Could we maybe leave `VMRegionWrapAround` or change it to `VMRegionOutOfRange` on line 164? Sorry about the back and forth...

Reply:
@mtoohey31 Changed in commit 90e94ff4881bca3ef9f629a8c70d286a75157db7 :)